mrjob/yarnemr.py

```diff
@@ -175,18 +176,32 @@ def _launch_yarn_emr_job(self):
                 _ResourceConstraint(operator.lt, 'appsRunning',
                                     _MAX_APPS_RUNNING)
             ]
-            cluster_id = self._find_cluster(resource_constraints)
-            if cluster_id:
-                self._cluster_id = cluster_id
+            find_cluster_ret = self._find_cluster(resource_constraints)
+            if isinstance(find_cluster_ret, str):
+                self._cluster_id = find_cluster_ret
+            elif isinstance(find_cluster_ret, int):
+                num_valid_clusters = find_cluster_ret
+            else:
+                assert False, 'Invalid return from `_find_cluster`'
 
         # create a cluster if we're not already using an existing one
-        if not self._cluster_id:
+        if self._cluster_id:
+            log.info('Adding our job to existing cluster %s (%s)' %
+                     (self._cluster_id, self._address_of_master()))
+        else:
+            if self._opts['max_pool_cluster_count'] and \
+              num_valid_clusters >= self._opts['max_pool_cluster_count']:
+                # In this case we max number of clusters has been reached
+                reason = ('Current cluster count ({}) >= max cluster'
+                          ' count ({})'
+                          .format(num_valid_clusters,
+                                  self._opts['max_pool_cluster_count']))
+                raise StepFailedException(
```

Comment:
is it possible to just keep pool waiting here instead of failing the step?

Reply:
I'm conflicted here. I could put it back in the pool. Let me think about it.

Reply:
back in pool is better since we wouldn't get a bunch of failures, we would just wait a bit longer for the clusters to spin up and have spots